### PR TITLE
SONARJAVA-2864 Fix nullness annotations from Spring when jsr305 binary is missing

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: java-frontend/src/test/java/org/sonar/java/se/JavaCheckVerifier.java [hunk 1]
```diff
@@ -177,16 +177,20 @@ public static void verifyIssueOnFile(String filename, String message, JavaFileSc
     javaCheckVerifier.scanFile(filename, new JavaFileScanner[] {check});
   }
 
-  public void scanFile(String filename, JavaFileScanner[] checks) {
+  public static Collection<File> classpathFromDirectory(String jarsDirectory) {
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/JavaCheckVerifier.java]: there's no need to change this class : your test should include only the required dependency.

File: java-frontend/src/test/java/org/sonar/java/se/checks/ConditionAlwaysTrueOrFalseCheckTest.java [hunk 2]
```diff
@@ -34,6 +36,14 @@ public void test() {
     JavaCheckVerifier.verify("src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java", new ConditionalUnreachableCodeCheck(), new BooleanGratuitousExpressionsCheck());
   }
 
+  @Test
+  public void test_without_jsr305() {
+    Collection<File> classpath = JavaCheckVerifier.classpathFromDirectory("target/test-jars").stream()
+      .filter(file -> !(file.getName().startsWith("jsr305-") || file.getName().startsWith("annotations-")))
+      .collect(Collectors.toList());
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/checks/ConditionAlwaysTrueOrFalseCheckTest.java]: This test is not very robust : you should only include the class you need to have your test case happening.

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseWithoutJSR305.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package my.app;
+
+import org.springframework.web.context.request.RequestAttributes;
```
Comment (benzonico) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseWithoutJSR305.java]: Name of the file is a bit dodgy : this is only testing spring annotations via the rule and not the rule : I would not use the ConditionAlwaysTrueOrFalse in the name of the file.

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java [hunk 4]
```diff
@@ -1325,6 +1325,18 @@ private Object nullableMethod() {
...
 
+  private void annotationWithNullableAndNonnullReturn() {
+    Object foo = annotationWithNullableAndNonnull();
+    if(foo == null) { // compliant, nullable wins
+    }
+  }
+
+  @javax.annotation.Nullable
+  @javax.annotation.Nonnull
+  private Object annotationWithNullableAndNonnull() {
```
Comment (benzonico) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: With the other testcase is that one still relevant ?

Comment (alban-auzeill) [reply java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: Yes. It shows explicitly our decision when we face a Nullable/Nonnull conflict.

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseWithoutJSR305.java (hunk 3 above)
Comment (alban-auzeill) [reply java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseWithoutJSR305.java]: renamed

**Pull request merged**